radix-engine/src/kernel/kernel_callback_api.rs

```diff
@@ -2,137 +2,150 @@ use super::actor::Actor;
-    where
-        Y: KernelApi<Self>;
+#[derive(Debug)]
+pub enum ReadSubstateEvent<'a> {
+    End {
+        handle: LockHandle,
+        value: &'a IndexedScryptoValue,
+    },
+}
 
-    fn before_create_node<Y>(
-        node_id: &NodeId,
-        node_substates: &NodeSubstates,
-        api: &mut Y,
-    ) -> Result<(), RuntimeError>
-    where
-        Y: KernelApi<Self>;
+#[derive(Debug)]
+pub enum WriteSubstateEvent<'a> {
+    Start {
+        handle: LockHandle,
+        value: &'a IndexedScryptoValue,
+    },
+}
 
-    fn after_create_node<Y>(
-        node_id: &NodeId,
-        total_substate_size: usize,
-        store_access: &StoreAccessInfo,
-        api: &mut Y,
-    ) -> Result<(), RuntimeError>
-    where
-        Y: KernelApi<Self>;
+#[derive(Debug)]
+pub enum CloseSubstateEvent<'a> {
+    StoreAccess(&'a StoreAccess),
+    End(LockHandle),
+}
 
-    fn after_move_modules<Y>(
-        src_node_id: &NodeId,
-        src_partition_number: PartitionNumber,
-        dest_node_id: &NodeId,
-        dest_partition_number: PartitionNumber,
-        store_access: &StoreAccessInfo,
-        api: &mut Y,
-    ) -> Result<(), RuntimeError>
-    where
-        Y: KernelApi<Self>;
+#[derive(Debug)]
+pub enum SetSubstateEvent<'a> {
+    Start(&'a IndexedScryptoValue),
+    StoreAccess(&'a StoreAccess),
+}
 
-    fn before_open_substate<Y>(
-        node_id: &NodeId,
-        partition_num: &PartitionNumber,
-        substate_key: &SubstateKey,
-        flags: &LockFlags,
-        api: &mut Y,
-    ) -> Result<(), RuntimeError>
-    where
-        Y: KernelApi<Self>;
+#[derive(Debug)]
+pub enum RemoveSubstateEvent<'a> {
+    Start,
+    StoreAccess(&'a StoreAccess),
+}
 
-    fn after_open_substate<Y>(
-        handle: LockHandle,
-        node_id: &NodeId,
-        size: usize,
-        store_access: &StoreAccessInfo,
-        api: &mut Y,
-    ) -> Result<(), RuntimeError>
+#[derive(Debug)]
+pub enum ScanKeysEvent<'a> {
+    Start,
+    StoreAccess(&'a StoreAccess),
+}
+
+#[derive(Debug)]
+pub enum DrainSubstatesEvent<'a> {
+    Start,
+    StoreAccess(&'a StoreAccess),
+}
+
+#[derive(Debug)]
+pub enum ScanSortedSubstatesEvent<'a> {
+    Start,
+    StoreAccess(&'a StoreAccess),
+}
+
+pub trait KernelCallbackObject: Sized {
+    type LockData: Default + Clone;
+
+    fn on_init<Y>(api: &mut Y) -> Result<(), RuntimeError>
     where
         Y: KernelApi<Self>;
 
-    fn after_read_substate<Y>(
-        lock_handle: LockHandle,
-        value_size: usize,
-        store_access: &StoreAccessInfo,
-        api: &mut Y,
-    ) -> Result<(), RuntimeError>
+    fn on_teardown<Y>(api: &mut Y) -> Result<(), RuntimeError>
     where
         Y: KernelApi<Self>;
 
-    fn after_write_substate<Y>(
-        lock_handle: LockHandle,
-        value_size: usize,
-        store_access: &StoreAccessInfo,
-        api: &mut Y,
-    ) -> Result<(), RuntimeError>
+    fn on_create_node<Y>(api: &mut Y, event: CreateNodeEvent) -> Result<(), RuntimeError>
```

Comment:
As discussed, let's revert to original code. It doesn't reduce complexity, and forces every handler to deal with enums. Can be done in another PR.